src/server/tiering/small_bins.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/small_bins.h"
+
+#include <algorithm>
+#include <optional>
+#include <utility>
+
+#include "absl/base/internal/endian.h"
+#include "base/logging.h"
+#include "core/compact_object.h"
+#include "server/tiering/disk_storage.h"
+
+namespace dfly::tiering {
+
+std::optional<SmallBins::FilledBin> SmallBins::Stash(std::string_view key, std::string_view value) {
+  DCHECK_LT(value.size(), 2_KB);
+
+  size_t value_bytes = 8 /* hash */ + 2 /* length */ + value.size();
+
+  std::optional<FilledBin> filled_bin;
+  if (current_bin_bytes_ + value_bytes >= 4_KB) {
+    filled_bin = FlushBin();
+  }
+
+  current_bin_bytes_ += value_bytes;
+  current_bin_[key] = value;
+  return filled_bin;
+}
+
+SmallBins::FilledBin SmallBins::FlushBin() {
+  std::string out;
+  out.resize(current_bin_bytes_ + 1);
+
+  unsigned id = ++last_bin_id_;
+  auto& pending_list = pending_bins_[id];
+
```

Comment:
add comments explaining how the layout looks like

Reply:
Added